cranelift/wasm/src/state.rs

```diff
@@ -241,6 +244,60 @@ pub struct FuncTranslationState {
     // `FuncEnvironment::make_direct_func()`.
     // Stores both the function reference and the number of WebAssembly arguments
     functions: HashMap<FuncIndex, (ir::FuncRef, usize)>,
+
+    // In each (C) function, stack arrays are allocated contiguously. We want
+    // to deterministically guarantee that no two consecutive stack allocations
+    // share the same MTE tag. Therefore, we generate an initial MTE tag
+    // randomly, and, consequently, increment that value (and apply modulo of
+    // course) and use that as the tag.
+    // We store the whole tagged index, which includes the MTE tag in its upper
+    // bits, instead of just the MTE tag.
+    latest_tagged_index: Option<Value>,
+}
+
+const MTE_NON_TAG_BITS_MASK: i64 = 0xF0FF_FFFF_FFFF_FFFFu64 as i64;
+const MTE_TAG_BITS_MASK: i64 = 0xF << 56;
+const MTE_TAG_ONE: i64 = 0x1 << 56;
+
+impl FuncTranslationState {
+    /// Tag the index with an MTE tag. If an initial tag has already been
+    /// generated, increment and use that, otherwise generate a new random tag
+    /// with IRG. Returns tagged index.
+    pub fn tag_index(&mut self, index: Value, builder: &mut FunctionBuilder) -> Value {
+        let tagged_index = match self.latest_tagged_index.take() {
+            None => {
+                // Generate random tag with IRG
+                let tagged_index = builder.ins().arm64_irg(index);
+
+                let var = Variable::new(0);
+                builder.declare_var(var, I64);
+                // let var = builder.def_var(b, val)
+                // builder.use_var(var)
+
+                tagged_index
+            }
+            Some(tagged_index) => {
+                // Extract random tag
+                let latest_tag = builder.ins().band_imm(tagged_index, MTE_TAG_BITS_MASK);
```

Comment:
This line currently panics on `band_imm`, if you follow the execution path further you will see that `value_type` gets called and the array access there panics because the `Value` doesn't exist.

Reply:
I tested the code without the variable declaration and it worked fine for me.

Reply:
Interesting. What input C code did you use for testing?